### PR TITLE
Feature/cassandra ha test

### PR DESCRIPTION
**Pull request by gmaretic**
---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1289)

### CODE REVIEW

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownDeleteTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+public class OneNodeDownDeleteTest {
+
+    @Rule
+    public ExpectedException expectException = ExpectedException.none();
+
+    @Test
+    public void deletingThrows() {
+        expectException.expect(PalantirRuntimeException.class);
+        OneNodeDownTestSuite.db.delete(OneNodeDownTestSuite.TEST_TABLE,
```
Comment (harrybiddle) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownDeleteTest.java]: Elsewhere we do this with 

```java
import static org.assertj.core.api.Assertions.assertThatThrownBy;

...

        assertThatThrownBy(() -> myFunction(myArgument)))
                .isInstanceOf(MyException.class);
```

Apparently the output is a bit nicer from this.

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Test
+    public void canGetRows() {
+        Map<Cell, Value> row = OneNodeDownTestSuite.db.getRows(OneNodeDownTestSuite.TEST_TABLE,
+                ImmutableList.of(OneNodeDownTestSuite.FIRST_ROW),
+                ColumnSelection.all(), Long.MAX_VALUE);
+
+        assertEquals(OneNodeDownTestSuite.DEFAULT_TIMESTAMP, row.get(OneNodeDownTestSuite.CELL_1_1).getTimestamp());
+        assertEquals(new String(OneNodeDownTestSuite.DEFAULT_VALUE),
```
Comment (harrybiddle) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java]: This is a bit tricky to read. I'd suggest storing the row you expect on OneNodeDownTestSuite (so that it lives next to the place where you build the row), and then check using assertj:

```java
import static org.assertj.core.api.Assertions.assertThat;

assertThat(row).containsAllEntriesOf(OneNodeDownTestSuite.FIRST_ROW_OF_TEST_TABLE);
```

It looks like `Value` already handles comparing the content byte arrays, so you wouldn't have to worry about making strings.

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                new String(row.get(OneNodeDownTestSuite.CELL_1_2).getContents()));
+    }
+
+    @Test
+    public void canGetRange() {
+        final RangeRequest range = RangeRequest.builder().endRowExclusive(OneNodeDownTestSuite.SECOND_ROW).build();
+        ClosableIterator<RowResult<Value>> it = OneNodeDownTestSuite.db.getRange(OneNodeDownTestSuite.TEST_TABLE, range,
+                Long.MAX_VALUE);
+
+        RowResult<Value> row = it.next();
```
Comment (harrybiddle) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java]: I'd construct an expected RowResult:

```
RowResult<Value> expectedRow = RowResult.of(...);
assertThat(it).containsOnly(expectedRow);
```

Here and throughout: you can use the iterable assertions of assertj (see [examples](https://github.com/joel-costigliola/assertj-examples/blob/master/assertions-examples/src/test/java/org/assertj/examples/IterableAssertionsExamples.java)), and rely on the equality methods of `Value` rather than extracting the values, constructing strings and comparing them, etc.

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+
+    @Test
+    public void canGetAllTimestamps() {
+        Multimap<Cell, Long> result = OneNodeDownTestSuite.db.getAllTimestamps(OneNodeDownTestSuite.TEST_TABLE,
+                ImmutableSet.of(OneNodeDownTestSuite.CELL_1_1), Long.MAX_VALUE);
+        assertEquals(2, result.get(OneNodeDownTestSuite.CELL_1_1).size());
+    }
+
+    protected static void verifyTimestampAndValue(Cell cell, long timestamp, byte[] value) {
```
Comment (harrybiddle) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java]: I think this would make more sense to be on `OneNodeDownTestSuite`

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    @BeforeClass
+    public static void initializeKvsAndDegradeCluster() throws IOException, InterruptedException {
+        setupTestTable();
+        degradeCassandraCluster();
+        db = createCassandraKvs();
+    }
+
+    @AfterClass
+    public static void bringClusterBack() throws IOException, InterruptedException {
```
Comment (harrybiddle) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java]: Seems badly named! Any reason why you got rid of the code to bring the node back up?

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        db.close();
+    }
+
+    private static void setupTestTable() {
+        db = createCassandraKvs();
```
Comment (harrybiddle) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java]: Suggest using a local variable rather than manipulating `db`. The fewer side effects of a method, the better: you've made the class-order dependent here.

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        // startup checks aren't guaranteed to pass immediately after killing the node, so we wait until
+        // they do. unclear if this is an AtlasDB product problem. see #1154
+        waitUntilStartupChecksPass();
+    }
+
+    private static void killFirstCassandraNode() throws IOException, InterruptedException {
+        Container container = CONTAINERS.getContainer(CASSANDRA_NODE_TO_KILL);
+        container.kill();
+    }
+
```
Comment (harrybiddle) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java]: Nit: double new lines here and elsewhere

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownDeleteTest.java (hunk 1 above)
Comment (gmaretic) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownDeleteTest.java]: Done

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java (hunk 2 above)
Comment (gmaretic) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java]: Done

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java (hunk 3 above)
Comment (gmaretic) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java]: Done

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java (hunk 4 above)
Comment (gmaretic) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java]: Done

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java (hunk 5 above)
Comment (gmaretic) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java]: It was adding extra time to the test. Do you think it makes sense to have it there anyway?

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java (hunk 6 above)
Comment (gmaretic) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownTestSuite.java]: Fixed

File: atlasdb-cassandra-multinode-tests/build.gradle [hunk 8]
```diff
@@ -14,4 +14,10 @@ dependencies {
 
     testCompile 'com.palantir.docker.compose:docker-compose-rule:' + libVersions.dockerComposeRule
     testCompile group: 'com.google.code.findbugs', name: 'annotations'
+    testCompile group: 'org.assertj', name: 'assertj-core', version: libVersions.assertj
```
Comment (harrybiddle) [inline atlasdb-cassandra-multinode-tests/build.gradle]: no need for `version` here: it's picked up from `versions.props`

```
testCompile group: 'org.assertj', name: 'assertj-core'
```

File: docs/source/release_notes/release-notes.rst [hunk 9]
```diff
@@ -41,6 +41,10 @@ develop
     *    - Type
          - Change
 
+    *    - |fixed|
+         - Multinode Cassandra: ``getAllTimestamps`` and ``getRangeOfTimestamps`` now work as long as quorum for reading is achieved.
```
Comment (harrybiddle) [inline docs/source/release_notes/release-notes.rst]: How about "`getAllTimestamps` and `getRangeOfTimestamps` now work on a multinode Cassandra cluster with one down node"?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 10]
```diff
@@ -1290,32 +1290,29 @@ public void dropTables(final Set<TableReference> tablesToDrop) {
 
     private void dropTablesInternal(final Set<TableReference> tablesToDrop) throws Exception {
         try {
-            clientPool.runWithRetry(new FunctionCheckedException<Client, Void, Exception>() {
```
Comment (harrybiddle) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: Why are we making a change to `dropTables` in this PR? From the release notes, I thought the only change was to the timestamp methods?

Comment (harrybiddle) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: Ahh, just prettification!

**Comment by jboreiko** [comment]
There are quite a few consistency changes in this PR.  We are aware that one is unacceptable when fetching all timestamps for a given cell (ensures sweep is able to detect all data in the cluster even if a partial delete has been processed).  All consistency changes should be discussed as a team.  I suspect we may have others that exist for a particular reason.

**Comment by ashrayjain** [comment]
big +1 to discuss the consistency changes as a team

**Comment by gmaretic** [comment]
Makes sense. FWIW, I did talk about this several times with the team, but will communicate better in the future.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java (hunk 10 above)
Comment (jboreiko) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: Generally this should be avoided in a PR with changes.  Makes it significantly harder to parse.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 11]
```diff
@@ -1210,13 +1212,13 @@ protected CfDef getCfForTable(TableReference tableRef, byte[] rawMetadata) {
                     rangeRequest,
                     timestampsGetterBatchSize.get(),
                     timestamp,
-                    deleteConsistency);
+                    readConsistency);
```
Comment (jboreiko) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: Will be reverted as discussed.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 12]
```diff
@@ -1670,7 +1669,7 @@ public void addGarbageCollectionSentinelValues(TableReference tableRef, Set<Cell
     @Override
     public Multimap<Cell, Long> getAllTimestamps(TableReference tableRef, Set<Cell> cells, long ts) {
         AllTimestampsCollector collector = new AllTimestampsCollector();
-        loadWithTs(tableRef, cells, ts, true, collector, deleteConsistency);
+        loadWithTs(tableRef, cells, ts, true, collector, readConsistency);
```
Comment (jboreiko) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: Will be reverted as discussed.

File: atlasdb-commons/src/main/java/com/palantir/common/base/Throwables.java [hunk 13]
```diff
@@ -82,7 +82,8 @@ public static RuntimeException rewrapAndThrowUncheckedException(String newMessag
      * If Throwable is a RuntimeException or Error, rethrow it. If not, throw a
      * new PalantirRuntimeException(ex)
      */
-    public static RuntimeException throwUncheckedException(Throwable ex) {
+    public static RuntimeException
```
Comment (jboreiko) [inline atlasdb-commons/src/main/java/com/palantir/common/base/Throwables.java]: This formatting change doesn't appear related to this PR.  Nor does it appear in line with our regular formatting.

File: docs/source/release_notes/release-notes.rst (hunk 9 above)
Comment (jboreiko) [inline docs/source/release_notes/release-notes.rst]: This will need to be reverted as well.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 14]
```diff
@@ -1062,7 +1062,7 @@ public String toString() {
                     }
                 });
             } catch (UnavailableException e) {
-                throw new PalantirRuntimeException("Truncating tables requires all Cassandra nodes"
+                throw new IllegalStateException("Truncating tables requires all Cassandra nodes"
```
Comment (jboreiko) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: Should be instead changed to an InsufficientConsistencyException

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 15]
```diff
@@ -1164,6 +1164,8 @@ public String toString() {
                             + numVersions + " total versions of " + cellVersionsMap.size() + " keys)";
                 }
             });
+        } catch (UnavailableException e) {
```
Comment (jboreiko) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: This will be an api break
